Dockerfile

```diff
@@ -1,100 +1,79 @@
 # This dockerfile expects proxies to be set via --build-arg if needed
 # It also expects to be contained in the /navigation2 root folder for file copy
 # Example build command:
-# sudo docker build -t nav2:latest --build-arg http_proxy=http://my.proxy.com:### .
-FROM osrf/ros2:bouncy-desktop
-
-SHELL ["/bin/bash", "-c"]
+# export http_proxy=http://my.proxy.com:80
+# export CMAKE_BUILD_TYPE=Debug
+# docker build -t nav2:latest --build-arg http_proxy --build-arg CMAKE_BUILD_TYPE ./
+FROM osrf/ros2:nightly
 
 # setup keys
-# check if proxy is set and get keys, using proxy if it is set
-RUN if [ "$http_proxy" == "" ]; \
+ARG http_proxy
+RUN if [ "$http_proxy" != "" ]; \
     then \
-      apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 \
-      --recv-keys 421C365BD9FF1F717815A3895523BAEEB01FA116; \
-    else \
       apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 \
       --keyserver-options http-proxy=$http_proxy \
       --recv-keys 421C365BD9FF1F717815A3895523BAEEB01FA116; \
     fi
 
-ENV ROS1_DISTRO melodic
-ENV ROS2_DISTRO bouncy
-ENV ROSDISTRO_INDEX_URL 'https://raw.githubusercontent.com/ros/rosdistro/master/index.yaml'
-
-# update latest package versions
-RUN apt-get update
-RUN rm -rf /etc/ros/rosdep/sources.list.d/20-default.list
-
 # install ROS2 dependencies
-RUN apt install -y \
-    build-essential \
-    cmake \
-    git \
-    python3-colcon-common-extensions \
-    python3-pip \
-    python-rosdep \
-    python3-vcstool \
-    wget
+RUN apt-get update && apt-get install -q -y \
+      build-essential \
+      cmake \
+      git \
+      python3-colcon-common-extensions \
+      python3-vcstool \
+      wget \
+    && rm -rf /var/lib/apt/lists/*
 
-# install Fast-RTPS dependencies
-RUN apt install --no-install-recommends -y \
-    libasio-dev \
-    libtinyxml2-dev
+# copy ros package repo
+ENV NAV2_WS /opt/nav2_ws
+RUN mkdir -p $NAV2_WS/src
+WORKDIR $NAV2_WS/src
+COPY ./ navigation2/
```

Comment:
This will make it so the dockerfile can't be run as a stand-alone outside of the navigation2 folder. I believe @SteveMacenski had a use case which needed that capability

Reply:
Most Docker enabled pipelines (DockerHub/Travis/CircleCI/GitLab) work on the premise of downloading the target Dockerfile and its build context before building, so deviating this pattern would make things harder to robustly integrate. For one, it would make it hard to ensure a one-to-one commit-to-ci relationship when the Dockerfile build is compiling the head of PR branch, and not the exact commit that triggered the CI 30-min ago while waiting in the Travis job/slot que. Building an image to test a branch using a Dockerfile not from that branch would be dangerous; like risking reproducibility by building from both the wrong blueprint and with outdated measurements.

Citing "Best practices for getting code into a container (git clone vs. copy vs. data container)":
> We never use git clone, beacuse using it you are not able to pull specific commit with easy way in CI and you are not interested in latest commit, cause during build on CI new commit can appear.

https://forums.docker.com/t/best-practices-for-getting-code-into-a-container-git-clone-vs-copy-vs-data-container/4077/3

Also, the previous Dockerfile from before was copying artifacts from the local build context, so checkout-and-copy was already necessary to get the context (if not the Dockerfile itself). Plus, it's not much of an issue if we want to clone a PR branch on github if we want to build the Dockerfile locally, and still be confident about the exact commit we want to test:
https://help.github.com/articles/checking-out-pull-requests-locally